tensorboard/plugins/debugger/interactive_debugger_plugin.py

```diff
@@ -0,0 +1,252 @@
+      # items are (node_name, output_slot, debug_op_name).
+      return http_util.Respond(request, {
+          'gated_grpc_tensors': gated,
+          'breakpoints': self._debugger_data_server.breakpoints,
+          'device_names': list(debug_graph_defs.keys()),
+      }, 'application/json')
+    elif mode == 'breakpoints':
+      # Retrieve currently enabled breakpoints.
+      return http_util.Respond(
+          request, self._debugger_data_server.breakpoints, 'application/json')
+    elif mode == 'set_state':
+      # Set the state of gated-gRPC debug tensors, e.g., disable, enable
+      # breakpoint.
+      node_name = request.args.get('node_name')
+      output_slot = int(request.args.get('output_slot'))
+      debug_op = request.args.get('debug_op')
+      state = request.args.get('state')
+      tf.logging.info('Setting state of %s:%d:%s to: %s' %
+                      (node_name, output_slot, debug_op, state))
+      if state == 'disable':
+        self._debugger_data_server.request_unwatch(
+            node_name, output_slot, debug_op)
+      elif state == 'watch':
+        self._debugger_data_server.request_watch(
+            node_name, output_slot, debug_op, breakpoint=False)
+      elif state == 'break':
+        self._debugger_data_server.request_watch(
+            node_name, output_slot, debug_op, breakpoint=True)
+      else:
+        tf.logging.error('Unrecognized new state for %s:%d:%s: %s' %
+                         (node_name, output_slot, debug_op, state))
+      return http_util.Respond(
+          request,
+          {'node_name': node_name,
+           'output_slot': output_slot,
+           'debug_op': debug_op,
+           'state': state},
+          'application/json')
+    else:
+      tf.logging.error('Unrecognized mode for the gated_grpc route: %s' % mode)
+
+  @wrappers.Request.application
+  def _serve_debugger_grpc_host_port(self, request):
+    return http_util.Respond(
+        request,
+        {'host': platform.node(), 'port': self._grpc_port}, 'application/json')
+
+  @wrappers.Request.application
+  def _serve_tensor_data(self, request):
+    response_encoding = 'application/json'
+    watch_key = request.args.get('watch_key')
+    time_indices = request.args.get('time_indices')
+    mapping = request.args.get('mapping')
+    slicing = request.args.get('slicing')
+
+    try:
+      sliced_tensor_data = self._debugger_data_server.query_tensor_store(
+          watch_key, time_indices=time_indices, slicing=slicing,
+          mapping=mapping)
+      response = {
+          'tensor_data': sliced_tensor_data,
+          'error': None
+      }
+      return http_util.Respond(request, response, response_encoding)
+    except (IndexError, ValueError) as e:
+      response = {
+          'tensor_data': None,
+          'error': {
+              'type': type(e).__name__,
+              'message': str(e),
+          },
+      }
+      return http_util.Respond(request, response, response_encoding)
```

Comment:
In cases like these, the HTTP status might want to reflect the error state too. If our TypeScript XHR stuff doesn't let you read the JSON of a 400 response, that's a bug we need to fix.